### PR TITLE
add full support for clLinkProgram

### PR DESCRIPTION
**Pull request by rjodinchr**
Use LLVM IR as the intermediate language
clCompileProgram generated a LLVM IR binary.
clLinkProgram takes LLVM IR binaries as inputs and generates a SPIRV binary.
cvk_program stores the LLVM IR in the new argument 'm_ir'.

Based on google/clspv#932
Ref #391

### CODE REVIEW

**Comment by rjodinchr** [comment]
This is passing all 'test_compiler' of the CTS using a swiftshader backend (except 'pragma_unroll' which needs int8 support).
'features_macro' pass for CL1.2, but not for CL3.0, unrelated to link programs.

I'll add some more tests in clvk soon in this PR.

**Comment by rjodinchr** [comment]
This PR is ready for review, but it needs https://github.com/google/clspv/pull/932.
I'll leave it as a draft for the time being.

File: src/program.hpp [hunk 1]
```diff
@@ -539,20 +539,124 @@ struct cvk_program : public _cl_program, api_object<object_magic::program> {
...
+        }
+        if (version != clvk_binary_version) {
+            cvk_warn_fn("wrong version");
+            return CL_PROGRAM_BINARY_TYPE_NONE;
+        }
+        return binary_type;
+    }
+
+public:
     CHECK_RETURN bool read(const unsigned char* src, size_t size) {
```
Comment (kpet) [inline src/program.hpp:598]: Maybe this function and a couple of others are a bit big for the header file. I know there are existing big-ish functions in headers but I'd like to change this where there are no performance or other considerations.

File: tests/azure/steps-run-tests.yml [hunk 2]
```diff
@@ -47,6 +47,15 @@ steps:
...
       condition: ${{ parameters.compilerAvailable }}
+    - script: |
+        set -ex
+        ./build/clspv -o simple.bc --emit-binary-ir ./tests/simple-from-llvm-ir-binary/simple.cl
+        export VK_ICD_FILENAMES=${{ parameters.icd_json_unix }}
+        export CLVK_LOG=3
+        ./build/simple_test_from_llvm_ir_binary
+        ./build/simple_test_from_llvm_ir_binary_static
+      displayName: Offline IR compilation simple tests
+      condition: ${{ parameters.compilerAvailable }}
```
Comment (kpet) [inline tests/azure/steps-run-tests.yml:58]: Wrong CI files :). We've switched to Github Actions months ago but I hadn't removed the Azure Pipelines files. I'm doing this with https://github.com/kpet/clvk/pull/431. Sorry for the confusion.

**Review by kpet — COMMENTED** [review]
Thanks! A couple comments after a cursory first look.

**Comment by rjodinchr** [comment]
I have divided this PR into 3 parts. This one, #432 and https://github.com/rjodinchr/clvk/pull/1 (I'll submit it once this one has been merged)

File: src/program.hpp [hunk 3]
```diff
@@ -653,6 +664,11 @@ struct cvk_program : public _cl_program, api_object<object_magic::program> {
     std::mutex m_lock;
     std::unique_ptr<std::thread> m_thread;
     std::string m_source;
+#ifndef CLSPV_ONLINE_COMPILER
+    std::vector<uint8_t> m_ir;
+#else
+    std::string m_ir;
+#endif
```
Comment (kpet) [inline src/program.hpp:671]: I don't think we need two different definitions here. Maybe handling vectors in the file saving utilities would be easier?

File: src/program.cpp [hunk 4]
```diff
@@ -828,77 +829,124 @@ std::string cvk_program::prepare_build_options(const cvk_device* device) const {
...
+            cmd += clspv_input_file;
+            cmd += " ";
         }
     }
 
-    // Select operation
-    // TODO support building a library with clBuildProgram
-    if (m_operation == build_operation::compile) {
-        m_binary_type = CL_PROGRAM_BINARY_TYPE_COMPILED_OBJECT;
+    std::string spirv_file{tmp_folder + "/compiled.spv"};
```
Comment (kpet) [inline src/program.cpp:911]: We should probably change the name depending on the output format.

File: src/program.cpp [hunk 5]
```diff
@@ -962,142 +987,126 @@ cl_build_status cvk_program::compile_source(const cvk_device* device) {
...
+        } else {
+            if (m_source.empty() && !m_ir.empty()) {
+                programs.push_back(m_ir);
+            } else {
+                programs.push_back(m_source);
+            }
+        }
+        m_build_log.clear();
+        if (build_to_ir) {
+            std::vector<uint32_t> ir;
```
Comment (kpet) [inline src/program.cpp:1010]: Looking at the whole chain from clspv to here, there are lots of conversions between uin32_t, steams, strings, etc. We should look at streamlining this a bit more.

File: src/program.hpp [hunk 6]
```diff
@@ -634,8 +634,19 @@ struct cvk_program : public _cl_program, api_object<object_magic::program> {
...
+#if COMPILER_AVAILABLE
+#ifndef CLSPV_ONLINE_COMPILER
+    CHECK_RETURN cl_build_status
+    build_source_offline(bool build_to_ir, bool build_from_il,
+                         std::string& build_options, std::string& tmp_folder);
+#else
+    CHECK_RETURN cl_build_status build_source_online(
+        bool build_to_ir, bool build_from_il, std::string& build_options);
+#endif
+#endif
```
Comment (kpet) [inline src/program.hpp:648]: I like how you split what was becoming a monster into online and offline functions but I don't think `build_source*` is a good name any more (the naming problem existed before this change).

**Review by kpet — APPROVED** [review]
This could be polished a bit more but, in the grand scheme of things these are minor issues, this change is a really nice step forward in terms of functionality and it's been in review for a while already :).

File: src/program.cpp [hunk 7]
```diff
@@ -828,77 +829,124 @@ std::string cvk_program::prepare_build_options(const cvk_device* device) const {
...
+        cvk_error_fn("failed to compile the program");
+        return CL_BUILD_ERROR;
+    }
+
+    if (build_to_ir) {
+        std::ifstream stream(spirv_file, std::ios::in | std::ios::binary);
+        m_ir.assign((std::istreambuf_iterator<char>(stream)),
+                    std::istreambuf_iterator<char>());
+        cvk_info("Loaded IR binary from \"%s\", size = %zu", spirv_file.c_str(),
+                 spirv_file.size());
```
Comment (kpet) [inline src/program.cpp:928]: This is reporting the binary's size as the length of the name of the file it was loaded from :)

**Pull request merged**